File: cli/internal/ci/vendors.go
package ci

type vendorEnvs struct {
	Any []string
	All []string
}

// Vendor describes a CI/CD vendor execution environment
type Vendor struct {
	// Name is the name of the vendor
	Name string
	// Constant is the environment variable prefix used by the vendor
	Constant string
	// Env is one or many environment variables that can be used to quickly determine the vendor (using simple os.Getenv(env) check)
	Env vendorEnvs
	// EvalEnv is key/value map of environment variables that can be used to quickly determine the vendor
	EvalEnv map[string]string

	// The name of the environment variable that contains the current git sha
	ShaEnvVar string

	// The name of the environment variable that contains the current checked out branch
	BranchEnvVar string
}

// Vendors is a list of common CI/CD vendors (from https://github.com/watson/ci-info/blob/master/vendors.json)
var Vendors = []Vendor{
	{
		Name:     "Appcircle",
		Constant: "APPCIRCLE",
		Env:      vendorEnvs{Any: []string{"AC_APPCIRCLE"}},
	},
	{
		Name:     "AppVeyor",
		Constant: "APPVEYOR",
		Env:      vendorEnvs{Any: []string{"APPVEYOR"}},
	},
	{
		Name:     "AWS CodeBuild",
		Constant: "CODEBUILD",
		Env:      vendorEnvs{Any: []string{"CODEBUILD_BUILD_ARN"}},
	},
	{
		Name:     "Azure Pipelines",
		Constant: "AZURE_PIPELINES",
		Env:      vendorEnvs{Any: []string{"SYSTEM_TEAMFOUNDATIONCOLLECTIONURI"}},
	},
	{
		Name:     "Bamboo",
		Constant: "BAMBOO",
		Env:      vendorEnvs{Any: []string{"bamboo_planKey"}},
	},
	{
		Name:     "Bitbucket Pipelines",
		Constant: "BITBUCKET",
		Env:      vendorEnvs{Any: []string{"BITBUCKET_COMMIT"}},
	},
	{
		Name:     "Bitrise",
		Constant: "BITRISE",
		Env:      vendorEnvs{Any: []string{"BITRISE_IO"}},
	},
	{
		Name:     "Buddy",
		Constant: "BUDDY",
		Env:      vendorEnvs{Any: []string{"BUDDY_WORKSPACE_ID"}},
	},
	{
		Name:     "Buildkite",
		Constant: "BUILDKITE",
		Env:      vendorEnvs{Any: []string{"BUILDKITE"}},
	},
	{
		Name:     "CircleCI",
		Constant: "CIRCLE",
		Env:      vendorEnvs{Any: []string{"CIRCLECI"}},
	},
	{
		Name:     "Cirrus CI",
		Constant: "CIRRUS",
		Env:      vendorEnvs{Any: []string{"CIRRUS_CI"}},
	},
	{
		Name:     "Codefresh",
		Constant: "CODEFRESH",
		Env:      vendorEnvs{Any: []string{"CF_BUILD_ID"}},
	},
	{
		Name:     "Codemagic",
		Constant: "CODEMAGIC",
		Env:      vendorEnvs{Any: []string{"CM_BUILD_ID"}},
	},
	{
		Name:     "Codeship",
		Constant: "CODESHIP",
		EvalEnv: map[string]string{
			"CI_NAME": "codeship",
		},
	},
	{
		Name:     "Drone",
		Constant: "DRONE",
		Env:      vendorEnvs{Any: []string{"DRONE"}},
	},
	{
		Name:     "dsari",
		Constant: "DSARI",
		Env:      vendorEnvs{Any: []string{"DSARI"}},
	},
	{
		Name:     "Expo Application Services",
		Constant: "EAS",
		Env:      vendorEnvs{Any: []string{"EAS_BUILD"}},
	},
	{
		Name:         "GitHub Actions",
		Constant:     "GITHUB_ACTIONS",
		Env:          vendorEnvs{Any: []string{"GITHUB_ACTIONS"}},
		ShaEnvVar:    "GITHUB_SHA",
		BranchEnvVar: "GITHUB_REF_NAME",
	},
	{
		Name:     "GitLab CI",
		Constant: "GITLAB",
		Env:      vendorEnvs{Any: []string{"GITLAB_CI"}},
	},
	{
		Name:     "GoCD",
		Constant: "GOCD",
		Env:      vendorEnvs{Any: []string{"GO_PIPELINE_LABEL"}},
	},
	{
		Name:     "Google Cloud Build",
		Constant: "GOOGLE_CLOUD_BUILD",
		Env:      vendorEnvs{Any: []string{"BUILDER_OUTPUT"}},
	},
	{
		Name:     "LayerCI",
		Constant: "LAYERCI",
		Env:      vendorEnvs{Any: []string{"LAYERCI"}},
	},
	{
		Name:     "Gerrit",
		Constant: "GERRIT",
		Env:      vendorEnvs{Any: []string{"GERRIT_PROJECT"}},
	},
	{
		Name:     "Hudson",
		Constant: "HUDSON",
		Env:      vendorEnvs{Any: []string{"HUDSON"}},
	},
	{
		Name:     "Jenkins",
		Constant: "JENKINS",
		Env:      vendorEnvs{All: []string{"JENKINS_URL", "BUILD_ID"}},
	},
	{
		Name:     "Magnum CI",
		Constant: "MAGNUM",
		Env:      vendorEnvs{Any: []string{"MAGNUM"}},
	},
	{
		Name:     "Netlify CI",
		Constant: "NETLIFY",
		Env:      vendorEnvs{Any: []string{"NETLIFY"}},
	},
	{
		Name:     "Nevercode",
		Constant: "NEVERCODE",
		Env:      vendorEnvs{Any: []string{"NEVERCODE"}},
	},
	{
		Name:     "ReleaseHub",
		Constant: "RELEASEHUB",
		Env:      vendorEnvs{Any: []string{"RELEASE_BUILD_ID"}},
	},
	{
		Name:     "Render",
		Constant: "RENDER",
		Env:      vendorEnvs{Any: []string{"RENDER"}},
	},
	{
		Name:     "Sail CI",
		Constant: "SAIL",
		Env:      vendorEnvs{Any: []string{"SAILCI"}},
	},
	{
		Name:     "Screwdriver",
		Constant: "SCREWDRIVER",
		Env:      vendorEnvs{Any: []string{"SCREWDRIVER"}},
	},
	{
		Name:     "Semaphore",
		Constant: "SEMAPHORE",
		Env:      vendorEnvs{Any: []string{"SEMAPHORE"}},
	},
	{
		Name:     "Shippable",
		Constant: "SHIPPABLE",
		Env:      vendorEnvs{Any: []string{"SHIPPABLE"}},
	},
	{
		Name:     "Solano CI",
		Constant: "SOLANO",
		Env:      vendorEnvs{Any: []string{"TDDIUM"}},
	},
	{
		Name:     "Sourcehut",
		Constant: "SOURCEHUT",
		EvalEnv: map[string]string{
			"CI_NAME": "sourcehut",
		},
	},
	{
		Name:     "Strider CD",
		Constant: "STRIDER",
		Env:      vendorEnvs{Any: []string{"STRIDER"}},
	},
	{
		Name:     "TaskCluster",
		Constant: "TASKCLUSTER",
		Env:      vendorEnvs{All: []string{"TASK_ID", "RUN_ID"}},
	},
	{
		Name:     "TeamCity",
		Constant: "TEAMCITY",
		Env:      vendorEnvs{Any: []string{"TEAMCITY_VERSION"}},
	},
	{
		Name:     "Travis CI",
		Constant: "TRAVIS",
		Env:      vendorEnvs{Any: []string{"TRAVIS"}},
	},
	{
		Name:         "Vercel",
		Constant:     "VERCEL",
		Env:          vendorEnvs{Any: []string{"NOW_BUILDER", "VERCEL"}},
		ShaEnvVar:    "VERCEL_GIT_COMMIT_SHA",
		BranchEnvVar: "VERCEL_GIT_COMMIT_REF",
	},
	{
		Name:     "Visual Studio App Center",
		Constant: "APPCENTER",
		Env:      vendorEnvs{Any: []string{"APPCENTER"}},
	},
	{
		Name:     "Woodpecker",
		Constant: "WOODPECKER",
		EvalEnv: map[string]string{
			"CI": "woodpecker",
		},
	},
	{
		Name:     "Xcode Cloud",
		Constant: "XCODE_CLOUD",
		Env:      vendorEnvs{Any: []string{"CI_XCODE_PROJECT"}},
	},
	{
		Name:     "Xcode Server",
		Constant: "XCODE_SERVER",
		Env:      vendorEnvs{Any: []string{"XCS"}},
	},
}

File: crates/turborepo-ci/src/vendors.rs
use std::{collections::HashMap, fmt::Debug, sync::OnceLock};

use crate::vendor_behavior::VendorBehavior;

#[derive(Clone, Debug, PartialEq)]
pub struct VendorEnvs {
    pub(crate) any: Vec<&'static str>,
    pub(crate) all: Vec<&'static str>,
}

#[derive(Clone, Debug, PartialEq)]
#[allow(dead_code)]
pub struct Vendor {
    pub(crate) name: &'static str,
    pub constant: &'static str,
    pub(crate) env: VendorEnvs,
    pub(crate) eval_env: Option<HashMap<&'static str, &'static str>>,
    pub sha_env_var: Option<&'static str>,
    pub branch_env_var: Option<&'static str>,
    pub username_env_var: Option<&'static str>,
    pub behavior: Option<VendorBehavior>,
}

static VENDORS: OnceLock<[Vendor; 45]> = OnceLock::new();

pub(crate) fn get_vendors() -> &'static [Vendor] {
    VENDORS
        .get_or_init(|| {
            [
                Vendor {
                    name: "Appcircle",
                    constant: "APPCIRCLE",
                    env: VendorEnvs {
                        any: vec!["AC_APPCIRCLE"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "AppVeyor",
                    constant: "APPVEYOR",
                    env: VendorEnvs {
                        any: vec!["APPVEYOR"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "AWS CodeBuild",
                    constant: "CODEBUILD",
                    env: VendorEnvs {
                        any: vec!["CODEBUILD_BUILD_ARN"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Azure Pipelines",
                    constant: "AZURE_PIPELINES",
                    env: VendorEnvs {
                        any: vec!["SYSTEM_TEAMFOUNDATIONCOLLECTIONURI"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: Some(VendorBehavior::new(
                        |group_name| format!("##[group]{group_name}\r\n"),
                        |_| String::from("##[endgroup]\r\n"),
                    )),
                },
                Vendor {
                    name: "Bamboo",
                    constant: "BAMBOO",
                    env: VendorEnvs {
                        any: vec!["bamboo_planKey"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Bitbucket Pipelines",
                    constant: "BITBUCKET",
                    env: VendorEnvs {
                        any: vec!["BITBUCKET_COMMIT"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Bitrise",
                    constant: "BITRISE",
                    env: VendorEnvs {
                        any: vec!["BITRISE_IO"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Buddy",
                    constant: "BUDDY",
                    env: VendorEnvs {
                        any: vec!["BUDDY_WORKSPACE_ID"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Buildkite",
                    constant: "BUILDKITE",
                    env: VendorEnvs {
                        any: vec!["BUILDKITE"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "CircleCI",
                    constant: "CIRCLE",
                    env: VendorEnvs {
                        any: vec!["CIRCLECI"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Cirrus CI",
                    constant: "CIRRUS",
                    env: VendorEnvs {
                        any: vec!["CIRRUS_CI"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Codefresh",
                    constant: "CODEFRESH",
                    env: VendorEnvs {
                        any: vec!["CF_BUILD_ID"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Codemagic",
                    constant: "CODEMAGIC",
                    env: VendorEnvs {
                        any: vec!["CM_BUILD_ID"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Codeship",
                    constant: "CODESHIP",
                    env: VendorEnvs {
                        any: vec![],
                        all: vec![],
                    },
                    eval_env: Some({
                        let mut map = HashMap::new();
                        map.insert("CI_NAME", "codeship");
                        map
                    }),
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Drone",
                    constant: "DRONE",
                    env: VendorEnvs {
                        any: vec!["DRONE"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "dsari",
                    constant: "DSARI",
                    env: VendorEnvs {
                        any: vec!["DSARI"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Expo Application Services",
                    constant: "EAS",
                    env: VendorEnvs {
                        any: vec!["EAS_BUILD"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "GitHub Actions",
                    constant: "GITHUB_ACTIONS",
                    env: VendorEnvs {
                        any: vec!["GITHUB_ACTIONS"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: Some("GITHUB_SHA"),
                    branch_env_var: Some("GITHUB_REF_NAME"),
                    username_env_var: Some("GITHUB_ACTOR"),
                    behavior: Some(
                        VendorBehavior::new(
                            |group_name| format!("::group::{group_name}\n"),
                            |_| String::from("::endgroup::\n"),
                        )
                        .with_error(
                            |group_name| format!("\x1B[;31m{group_name}\x1B[;0m\n"),
                            |_| String::new(),
                        ),
                    ),
                },
                Vendor {
                    name: "GitLab CI",
                    constant: "GITLAB",
                    env: VendorEnvs {
                        any: vec!["GITLAB_CI"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "GoCD",
                    constant: "GOCD",
                    env: VendorEnvs {
                        any: vec!["GO_PIPELINE_LABEL"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Google Cloud Build",
                    constant: "GOOGLE_CLOUD_BUILD",
                    env: VendorEnvs {
                        any: vec!["BUILDER_OUTPUT"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "LayerCI",
                    constant: "LAYERCI",
                    env: VendorEnvs {
                        any: vec!["LAYERCI"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Gerrit",
                    constant: "GERRIT",
                    env: VendorEnvs {
                        any: vec!["GERRIT_PROJECT"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Hudson",
                    constant: "HUDSON",
                    env: VendorEnvs {
                        any: vec!["HUDSON"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Jenkins",
                    constant: "JENKINS",
                    env: VendorEnvs {
                        any: vec![],
                        all: vec!["JENKINS_URL", "BUILD_ID"],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Magnum CI",
                    constant: "MAGNUM",
                    env: VendorEnvs {
                        any: vec!["MAGNUM"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Netlify CI",
                    constant: "NETLIFY",
                    env: VendorEnvs {
                        any: vec!["NETLIFY"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Nevercode",
                    constant: "NEVERCODE",
                    env: VendorEnvs {
                        any: vec!["NEVERCODE"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "ReleaseHub",
                    constant: "RELEASEHUB",
                    env: VendorEnvs {
                        any: vec!["RELEASE_BUILD_ID"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Render",
                    constant: "RENDER",
                    env: VendorEnvs {
                        any: vec!["RENDER"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Sail CI",
                    constant: "SAIL",
                    env: VendorEnvs {
                        any: vec!["SAILCI"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Screwdriver",
                    constant: "SCREWDRIVER",
                    env: VendorEnvs {
                        any: vec!["SCREWDRIVER"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Semaphore",
                    constant: "SEMAPHORE",
                    env: VendorEnvs {
                        any: vec!["SEMAPHORE"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Shippable",
                    constant: "SHIPPABLE",
                    env: VendorEnvs {
                        any: vec!["SHIPPABLE"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Solano CI",
                    constant: "SOLANO",
                    env: VendorEnvs {
                        any: vec!["TDDIUM"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Sourcehut",
                    constant: "SOURCEHUT",
                    env: VendorEnvs {
                        any: vec![],
                        all: vec![],
                    },
                    eval_env: Some({
                        let mut map = HashMap::new();
                        map.insert("CI_NAME", "sourcehut");
                        map
                    }),
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Strider CD",
                    constant: "STRIDER",
                    env: VendorEnvs {
                        any: vec!["STRIDER"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "TaskCluster",
                    constant: "TASKCLUSTER",
                    env: VendorEnvs {
                        any: vec!["TASK_ID", "RUN_ID"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "TeamCity",
                    constant: "TEAMCITY",
                    env: VendorEnvs {
                        any: vec!["TEAMCITY_VERSION"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: Some(VendorBehavior::new(
                        |group_name| format!("##teamcity[blockOpened name='{group_name}']"),
                        |group_name| format!("##teamcity[blockClosed name='{group_name}']"),
                    )),
                },
                Vendor {
                    name: "Travis CI",
                    constant: "TRAVIS",
                    env: VendorEnvs {
                        any: vec!["TRAVIS"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: Some(VendorBehavior::new(
                        |group_name| format!("travis_fold:start:{group_name}\r\n"),
                        |group_name| format!("travis_fold:end:{group_name}\r\n"),
                    )),
                },
                Vendor {
                    name: "Vercel",
                    constant: "VERCEL",
                    env: VendorEnvs {
                        any: vec!["NOW_BUILDER", "VERCEL"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: Some("VERCEL_GIT_COMMIT_SHA"),
                    branch_env_var: Some("VERCEL_GIT_COMMIT_REF"),
                    username_env_var: Some("VERCEL_GIT_COMMIT_AUTHOR_LOGIN"),
                    behavior: None,
                },
                Vendor {
                    name: "Visual Studio App Center",
                    constant: "APPCENTER",
                    env: VendorEnvs {
                        any: vec!["APPCENTER"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Woodpecker",
                    constant: "WOODPECKER",
                    env: VendorEnvs {
                        any: vec![],
                        all: vec![],
                    },
                    eval_env: Some({
                        let mut map = HashMap::new();
                        map.insert("CI", "woodpecker");
                        map
                    }),
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Xcode Cloud",
                    constant: "XCODE_CLOUD",
                    env: VendorEnvs {
                        any: vec!["CI_XCODE_PROJECT"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
                Vendor {
                    name: "Xcode Server",
                    constant: "XCODE_SERVER",
                    env: VendorEnvs {
                        any: vec!["XCS"],
                        all: vec![],
                    },
                    eval_env: None,
                    sha_env_var: None,
                    branch_env_var: None,
                    username_env_var: None,
                    behavior: None,
                },
            ]
        })
        .as_slice()
}
